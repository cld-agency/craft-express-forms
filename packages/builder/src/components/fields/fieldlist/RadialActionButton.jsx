import PropTypes from 'prop-types';
import React, { useState } from 'react';
import translate from '../../../functions/translator';
import types from '../types';
import './radialActionButton.styl';
import text from './icons/fieldtype-text.svg';
import textarea from './icons/fieldtype-textarea.svg';
import checkbox from './icons/fieldtype-checkbox.svg';
import email from './icons/fieldtype-email.svg';
import file from './icons/fieldtype-file.svg';
import hidden from './icons/fieldtype-hidden.svg';
import options from './icons/fieldtype-options.svg';

const icons = { text, textarea, checkbox, email, file, hidden, options };

const RadialActionButton = ({ type, changeTypeHandler }) => {
  const [isOpen, setOpen] = useState(false);

  const getIcon = (type) => {
    const IconComponent = icons[type];

    return <IconComponent />;
  };

  return (
    <>
      {isOpen && <div className="radial-action-button-overlay" onClick={() => setOpen(false)} />}
      <div className={`radial-action-button${isOpen ? ' open' : ''}`}>
        <ul className={`items data-type-count-${types.length - 1}`}>
          {types.map(
            (itemType) =>
              itemType !== type && (
                <li key={itemType}>
                  <a
                    onClick={() => {
                      changeTypeHandler(itemType);
                      setOpen(false);
                    }}
                    data-field-type={itemType}
                    title={translate(itemType[0].toUpperCase() + itemType.slice(1))}
                    tabIndex={-1}
                  >
                    <span>{getIcon(itemType)}</span>
                  </a>
                </li>
              )
          )}
        </ul>

        <button className="action-button" tabIndex={-1} onClick={() => setOpen(true)}>
          <span className="type-icon">{getIcon(type)}</span>
        </button>

        <div className="close-button" onClick={() => setOpen(false)} />

        <div className="backdrop" />
      </div>
    </>
  );
};

RadialActionButton.propTypes = {
  type: PropTypes.oneOf(types).isRequired,
  changeTypeHandler: PropTypes.func,
};

export default RadialActionButton;
