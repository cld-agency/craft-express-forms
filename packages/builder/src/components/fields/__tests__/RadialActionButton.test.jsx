import React from 'react';
import { shallow } from 'enzyme';
import RadialActionButton from '../fieldlist/RadialActionButton';

describe('<RadialActionButton />', () => {
  it('renders & displays correctly', () => {
    const wrapper = shallow(<RadialActionButton type="text" />);
    expect(wrapper).toMatchSnapshot();
  });
});
