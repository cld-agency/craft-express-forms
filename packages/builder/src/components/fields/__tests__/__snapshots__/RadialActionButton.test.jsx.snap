// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadialActionButton /> renders & displays correctly 1`] = `
<Fragment>
  <div
    className="radial-action-button"
  >
    <ul
      className="items data-type-count-6"
    >
      <li
        key="textarea"
      >
        <a
          data-field-type="textarea"
          onClick={[Function]}
          tabIndex={-1}
          title="Textarea"
        >
          <span>
            <SVGElement />
          </span>
        </a>
      </li>
      <li
        key="options"
      >
        <a
          data-field-type="options"
          onClick={[Function]}
          tabIndex={-1}
          title="Options"
        >
          <span>
            <SVGElement />
          </span>
        </a>
      </li>
      <li
        key="checkbox"
      >
        <a
          data-field-type="checkbox"
          onClick={[Function]}
          tabIndex={-1}
          title="Checkbox"
        >
          <span>
            <SVGElement />
          </span>
        </a>
      </li>
      <li
        key="email"
      >
        <a
          data-field-type="email"
          onClick={[Function]}
          tabIndex={-1}
          title="Email"
        >
          <span>
            <SVGElement />
          </span>
        </a>
      </li>
      <li
        key="hidden"
      >
        <a
          data-field-type="hidden"
          onClick={[Function]}
          tabIndex={-1}
          title="Hidden"
        >
          <span>
            <SVGElement />
          </span>
        </a>
      </li>
      <li
        key="file"
      >
        <a
          data-field-type="file"
          onClick={[Function]}
          tabIndex={-1}
          title="File"
        >
          <span>
            <SVGElement />
          </span>
        </a>
      </li>
    </ul>
    <button
      className="action-button"
      onClick={[Function]}
      tabIndex={-1}
    >
      <span
        className="type-icon"
      >
        <SVGElement />
      </span>
    </button>
    <div
      className="close-button"
      onClick={[Function]}
    />
    <div
      className="backdrop"
    />
  </div>
</Fragment>
`;
